niworkflows/interfaces/bids.py

```diff
@@ -392,110 +401,114 @@ class DerivativesDataSink(SimpleInterface):
+            out_entities["extension"] = out_entities["extension"][0]
 
-        if self.inputs.space and '_res-native' in self.inputs.space:
-            # strip native tag
-            self.inputs.space = self.inputs.space.split('_')[0]
-
-        space = '_space-{}'.format(self.inputs.space) if self.inputs.space else ''
-        desc = '_desc-{}'.format(self.inputs.desc) if self.inputs.desc else ''
-        suffix = '_{}'.format(self.inputs.suffix) if self.inputs.suffix else ''
-        dtype = '' if not self.inputs.keep_dtype else ('_%s' % dtype)
+        # Insert custom (non-BIDS) entities from allowed_entities.
+        custom_entities = set(out_entities.keys()) - set(BIDS_DERIV_ENTITIES)
+        patterns = BIDS_DERIV_PATTERNS
+        if custom_entities:
+            custom_pat = "_".join(f"{key}-{{{key}}}" for key in sorted(custom_entities))
+            patterns = [pat.replace("_{suffix", "_".join(('', custom_pat, "{suffix")))
+                        for pat in patterns]
 
+        # Prepare SimpleInterface outputs object
+        self._results['out_file'] = []
         self._results['compression'] = []
-        self._results['fixed_hdr'] = [False] * len(self.inputs.in_file)
-
-        for i, fname in enumerate(self.inputs.in_file):
-            extra = ''
-            if isdefined(self.inputs.extra_values):
-                extra = '_{}'.format(self.inputs.extra_values[i])
-            out_file = formatstr.format(
-                bname=base_fname,
-                space=space,
-                desc=desc,
-                extra=extra,
-                suffix=suffix,
-                i=i,
-                dtype=dtype,
-                ext=ext,
-            )
-            self._results['out_file'].append(out_file)
-            self._results['compression'].append(_copy_any(fname, out_file))
-
-            is_nii = out_file.endswith(('.nii', '.nii.gz'))
+        self._results['fixed_hdr'] = [False] * len(in_file)
+
+        dest_files = build_path(out_entities, path_patterns=patterns)
+        if not dest_files:
+            raise ValueError(f"Could not build path with entities {out_entities}.")
+
+        # Make sure the interpolated values is embedded in a list, and check
+        dest_files = listify(dest_files)
+        if len(in_file) != len(dest_files):
+            raise ValueError(f"Input files ({len(in_file)}) not matched "
+                             f"by interpolated patterns ({len(dest_files)}).")
+
+        for i, (orig_file, dest_file) in enumerate(zip(in_file, dest_files)):
+            out_file = out_path / dest_file
+            out_file.parent.mkdir(exist_ok=True, parents=True)
+            self._results['out_file'].append(str(out_file))
+            self._results['compression'].append(_copy_any(orig_file, str(out_file)))
+
+            is_nifti = out_file.name.endswith(('.nii', '.nii.gz')) \
+                and not out_file.name.lstrip(".gz").endswith('.dtseries.nii')
```

Comment:
this was being checked afterward, anyways.

Reply:
is the `lstrip(".gz")` necessary? I don't think we generate compressed CIFTIs

Reply:
do you foresee any potential problems from having it? If now, then it's better having a more generalizable interface, WDTY?

Reply:
good catch. I might have missed the starting point somewhere else so I'll double check.

Reply:
I finally opted for `endswith((".dtseries.nii", ".dtseries.nii.gz"))` just in case. I guess checking for compressed cifti is not such a big price to pay.